# CSE_pipeline

Process to generate Chromatin State Expression (CSE) matrix

<img width="1176" alt="image" src="https://github.com/fw262/CSE_pipeline/assets/56937181/f7aa17b6-b402-4343-bbc7-19554d5146f4">

This pipeline takes in bam files from bulk RNA-seq data generated by the nf-core and generates a Chromatin State Expression matrix. The annotations are based on the full stack ChromHMM annotations from [here](https://github.com/ernstlab/full_stack_ChromHMM_annotations/tree/main).

### Required tools
1) [snakemake](https://snakemake.readthedocs.io/en/stable/)
2) [bedtools](https://bedtools.readthedocs.io/en/latest/content/installation.html)
3) [R](https://www.r-project.org/)
4) [MACS3](https://macs3-project.github.io/MACS/docs/INSTALL.html)
5) [Subread package, specifically featureCounts](https://github.com/ShiLab-Bioinformatics/subread)

### Instructions
1) Copy or link bam files to a folder called *data*.
2) List the sample names (no .bam suffix) as individual lines in the text file *sampleNames.txt*.
3) Run snakemake command

### Output
1) *DNX_data_count_anno.tsv* contains the counts for each chromatin state feature (rows) for each sample (columns).
2) *peakCountBySample_broadPeaks* folder contains the MACS3 peak calling outputs for each samples for both + and - strands.
3) *peakCountBySample_broadPeaks_gtf* folder contains peak annotations in gtf format.
4) *peakCountBySample_broadPeaks_featureCounts* contains counts for each peak feature, annotated with chromatin states, for each sample.
